backend/tests/test_auth.py

```diff
@@ -0,0 +1,75 @@
+import pytest
+from backend.database import User
+from backend.auth import user_manager
+from flask_jwt_extended import decode_token
+
+
+@pytest.mark.parametrize(
+    ("email", "password", "expected_status_code"),
+    [
+        ("test@email.com", "password", 200),
+        ("bad_email", "bad_password", 400),
+        (None, None, 400),
+    ],
+)
+def test_register(client, email, password, expected_status_code):
+    res = client.post(
+        "api/v1/auth/register",
+        json={
+            "email": email,
+            "password": password,
+        },
+    )
+    assert res.status_code == expected_status_code
+
+
+@pytest.mark.parametrize(
+    ("password", "expected_status_code"),
+    [("password", 200), ("bad_password", 401), (None, 400)],
+)
+def test_login(client, db_session, password, expected_status_code):
+    test_user = User(
+        email="test2@email.com",
+        password=user_manager.hash_password("password"),
+        first_name="first",
+        last_name="last",
+    )
+    db_session.add(test_user)
+    db_session.commit()
```

Comment:
You can create a fixture that creates a test user. The package `pytest-flask-sqlalchemy` rolls back the state of the db for each test so it's safe to continually recreate the same thing again and again.

Because we can guarantee the state of the database thanks to the aforementioned package and how we fire up the test database (reset it each test run), it also becomes much easier to test the case that an email _doesn't_ exist. So I would add `email` as a test parameter, and test for non-existing emails.

Finally, I would also use the string `"my_password"` instead of `"password"` because "my_password" should never occur anywhere in the code base other than in the tests. Protects against the possibility someone accidentally makes it so the code is checking for, literally, `"password"`.

The combination of all the above would be something like this. Haven't run this code but I think you can copy paste to replace from here to the end of the file

```python
@pytest.fixture
def example_user(db_session):
    user = User(
        email="test@email.com",
        password=user_manager.hash_password("my_password"),
        first_name="first",
        last_name="last",
    )
    db_session.add(user)
    db_session.commit()
    yield


@pytest.mark.parametrize(
    ("email", "password", "expected_status_code"),
    [
        ("test@email.com", "my_password", 200),
        ("test@email.com", "bad_password", 401),
        ("bad_email@email.com", "my_password", 401),
        ("test@email.com", None, 400)
    ],
)
def test_login(client, example_user, email, password, expected_status_code):
    res = client.post(
        "api/v1/auth/login",
        json={
            "email": email,
            "password": password,
        },
    )

    assert res.status_code == expected_status_code


 def test_jwt(client, example_user):
    res = client.post(
        "api/v1/auth/login",
        json={
            "email": "test@email.com",
            "password": "my_password",
        },
    )

    assert res.status_code == 200

    user_id = decode_token(res.json["access_token"])["sub"]
    db_user = db_session.query(User).filter(user_id == User.id).first()

    assert db_user.email == example_user.email
    assert res.status_code == 200
```

Reply:
Getting an error from the DB implementing this, seems like the DB might not be resetting between tests?

```
sqlalchemy.exc.IntegrityError: (psycopg2.errors.UniqueViolation) duplicate key value violates unique constraint "user_email_key"
E       DETAIL:  Key (email)=(test@email.com) already exists.
```

Reply:
Adding `db_session` as a parameter to the test that was causing this error fixed this problem

Reply:
excellent